clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -303,7 +303,14 @@ private void throwIfOffsetOutOfRange() throws OffsetOutOfRangeException {
 
             for (PartitionRecords<K, V> part : this.records) {
                 if (!subscriptions.isFetchable(part.partition)) {
-                    log.debug("Ignoring fetched records for {} since it is no longer fetchable", part.partition);
+                    // this can happen when a rebalance happened or a partition consumption paused
+                    // before fetched records are returned to the consumer's poll call
+                    log.debug("Not returning fetched records for partition {} since it is no longer fetchable", part.partition);
+
+                    // we also need to reset the fetch positions to pretend we did not fetch
+                    // this partition in the previous request at all
+                    subscriptions.fetched(part.partition, part.records.get(0).offset());
```

Comment:
@hachikuji I thought about this, but consumed() may return null for the first time so we cannot rely on that value either. Maybe we should just combine these two? I.e.:

If consumed() != null, set to consumed;
otherwise this MUST be the first fetch ever returned before any seek() was called, hence resetting to first offset should be safe.